### PR TITLE
prepare LXD environemnt (client + 2 controllers)

### PR DESCRIPTION
**Pull request by rgildein**
- creates "client" LXD container with JujuSpell installed
- bootstrap N controllers

TODO:
- [x] figure out why `setup_environment` works when each step is run individually, but not at once
- [x] figure out why `execute` raises [this](https://pastebin.ubuntu.com/p/2QcTSWCBSd/) errors (command is executed, but this error is always there)
- [x] creates config file for JujuSpell and copy paste it to the client
- [x] add simple test with ping comand

### CODE REVIEW

**Comment by rgildein** [comment]
Not included in this PR:

- [ ] README.md for functional tests with explanation
- [ ] configure GitHub workflow

File: tests/functional/conftest.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+DEFAULT_SERIES = "jammy"
+CLIENT_CONAINER = pytest.StashKey[str]()
+CONTROLLERS_CONTAINERS = pytest.StashKey[list]()
+
+
+def check_dependencies(*dependencies) -> None:
+    """Check if dependencies are installed."""
+    errors = []
+    for dependence in dependencies:
```
Comment (esunar) [inline tests/functional/conftest.py:40]: dependence -> dependency

File: tests/functional/setup_lxd.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+
+def cleanup_environment():
+    """Clean up LXD environment.
+
+    Remove all instances with names starting with CONTAINER_PREFIX.
+    """
+    client = Client()
+    for instance in client.instances.all():
+        if instance.name.startswith(CONTAINER_PREFIX):
```
Comment (esunar) [inline tests/functional/setup_lxd.py:208]: We should create a unique string for each run

**Review by esunar — CHANGES_REQUESTED** [review]

File: tests/functional/conftest.py (hunk 1 above)
Comment (rgildein) [reply tests/functional/conftest.py:40]: I think that bout are valid expressions. Do we really need to switch here?

**Comment by rgildein** [comment]
I added README.md and configured GH workflows in #52.

Examples of usage with `--no-build` and `--keep-env`.

File: tests/functional/conftest.py (hunk 1 above)
Comment (esunar) [reply tests/functional/conftest.py:42]: just because everyone uses dependency like dependency injection

File: tests/functional/setup_lxd.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    """Clean up LXD environment.
+
+    Remove all instances with names starting with CONTAINER_PREFIX.
+    """
+    client = Client()
+    for instance in client.instances.all():
+        if str(session_uuid) in instance.name:
+            if keep_env:
+                if (
+                    instance.status_code != 102
```
Comment (esunar) [inline tests/functional/setup_lxd.py:206]: can you define a constant for 102

File: tests/functional/test_connection.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import json
+
+
+def test_help(juju_spell_run):
+    """Test printing --help."""
+    help_flag = juju_spell_run("--help")
+    assert help_flag.exit_code == 0
+
+    help_default = juju_spell_run()
+    assert help_default.exit_code == 127
```
Comment (esunar) [inline tests/functional/test_connection.py:10]: can you define a constant for 127

File: tests/functional/setup_lxd.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    print(f"LXD: {container.name} is running")
+    return container
+
+
+def create_client_instance(
+    session_uuid: str, series: str, snap_path: Union[str, Path]
+) -> Instance:
+    """Create client instance."""
+    client = create_container(session_uuid, "client", series)
+    sleep(5)  # TODO: we need to wait for container to be fully readyx
```
Comment (esunar) [inline tests/functional/setup_lxd.py:90]: Random sleeps!! Can we make it a function to run a simple command on container until it succeeds?

**Review by esunar — CHANGES_REQUESTED** [review]

**Comment by esunar** [comment]
Can you also update Makefile

File: tests/functional/conftest.py (hunk 1 above)
Comment (jneo8) [reply tests/functional/conftest.py:42]: dependence is abstract whereas dependency is more concrete

File: tests/functional/setup_lxd.py (hunk 5 above)
Comment (rgildein) [reply tests/functional/setup_lxd.py:90]: I put TODO there and forgot to do anything about it. I'll create function.

**Review by esunar — APPROVED** [review]

**Pull request merged**